### PR TITLE
Finetuning NeMo Megatron T5 Models on GLUE

### PR DESCRIPTION
**Pull request by MaximumEntropy**

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_glue_model.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+    def setup_test_data(self):
+        self._test_dl = self.build_pretraining_data_loader(
+            self._test_ds,
+            self.cfg.data.test_ds.batch_size,
+            shuffle=False,
+            num_workers=self.cfg.data.test_ds.num_workers,
+            pin_memory=True,
+        )
+
```
Comment (yidong72) [inline nemo/collections/nlp/models/language_modeling/megatron_glue_model.py:267]: In your `megatron_t5_model.py` file, the `list_available_models` is class method, make it consistent?

File: nemo/collections/nlp/models/language_modeling/megatron_glue_model.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    Megatron T5 finetuning for GLUE
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg=cfg, trainer=trainer)
+        self.cfg = cfg
+        self.acc_metric = ExactStringMatchMetric()
+
+    def training_step(self, batch, batch_idx):
+        tokens_enc, tokens_dec, loss_mask, labels, enc_mask, dec_mask, enc_dec_mask = self.process_batch(batch)
```
Comment (yidong72) [inline nemo/collections/nlp/models/language_modeling/megatron_glue_model.py:106]: Is it really needed? Each process should get a copy of data from its dataloader. No need to `broadcast` right?

File: nemo/collections/nlp/data/glue_benchmark/glue_benchmark_dataset.py [hunk 3]
```diff
@@ -347,6 +358,118 @@ def _truncate_seq_pair(self, tokens_a: str, tokens_b: str, max_length: int):
...
+        Inputs are prefixed with a text prompt that indicates the task to perform.
+        """
+        features = []
+        for ex_index, example in enumerate(self.examples):
+            if ex_index % 10000 == 0:
+                logging.info(f"Writing example {ex_index} of {len(self.examples)}")
+
+            text_to_text_query = self.processor.get_t5_prompted_query(example.text_a, example.text_b)
+            enc_query = self.tokenizer.text_to_ids(text_to_text_query)
+            if len(enc_query) > self.max_seq_length:
```
Comment (yidong72) [inline nemo/collections/nlp/data/glue_benchmark/glue_benchmark_dataset.py:457]: The GLUE dataset restricts the examples to have `text_a`, `text_b` only where the T5 is more generic than that. We can use the dataset like the one in the GPT p-tuning PR, which can handle any fields in the data for different tasks.

File: examples/nlp/language_modeling/conf/megatron_t5_config_finetune_glue_mnli.yaml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    model_parallel_size: ${model.tensor_model_parallel_size}
+    save_best_model: True
+
+model:
+  restore_from_path: ??? # Path to a trained T5 .nemo file
+  tensor_model_parallel_size: 1
+
+  data:
+    train_ds:
+      task_name: 'mnli'
```
Comment (yidong72) [inline examples/nlp/language_modeling/conf/megatron_t5_config_finetune_glue_mnli.yaml:44]: I removed the `task_name` from the config. and added `task_name` to the data record so the model can handle multiple tasks simultaneously.

**Review by yidong72 — COMMENTED** [review]
Looks good overall. only a few tiny comments.

File: examples/nlp/language_modeling/conf/megatron_t5_config_finetune_glue_mnli.yaml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      file_path: '/raid/Data/GLUE/MNLI/train.tsv'
```
Comment (ekmb) [inline examples/nlp/language_modeling/conf/megatron_t5_config_finetune_glue_mnli.yaml:45]: set file_path to ???

File: examples/nlp/language_modeling/conf/megatron_t5_config_finetune_glue_mnli.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      batch_size: 32
+      shuffle: True
+      num_workers: 8
+      pin_memory: True
+      max_seq_length: 512
+
+    validation_ds:
+      task_name: 'mnli'
+      file_path: '/raid/Data/GLUE/MNLI/dev_matched.tsv'
```
Comment (ekmb) [inline examples/nlp/language_modeling/conf/megatron_t5_config_finetune_glue_mnli.yaml:54]: set file_path to ???

File: examples/nlp/language_modeling/megatron_t5_glue.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ekmb) [inline examples/nlp/language_modeling/megatron_t5_glue.py:1]: 2020-> 2022 + `with `affiliates` is missing. here and in other new files

File: nemo/collections/nlp/models/language_modeling/megatron_glue_model.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+            return
+        self.setup_training_data()
+
+    def setup_training_data(self):
+        self._train_dl = self.build_pretraining_data_loader(
+            self._train_ds,
+            self.cfg.data.train_ds.batch_size,
+            shuffle=True,
+            num_workers=self.cfg.data.train_ds.num_workers,
+            pin_memory=True,
```
Comment (ekmb) [inline nemo/collections/nlp/models/language_modeling/megatron_glue_model.py:257]: for setup_*_data methods shuffle and pin_memory are hardcoded not set via config?

**Review by ekmb — APPROVED** [review]
GLUE dataset-related changes LGTM.

File: nemo/collections/nlp/models/language_modeling/megatron_glue_model.py (hunk 2 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_glue_model.py]: Yeah, right now I'm just using process_batch to move things to torch tensors.

File: nemo/collections/nlp/models/language_modeling/megatron_glue_model.py (hunk 1 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_glue_model.py:269]: Fixed!

File: nemo/collections/nlp/models/language_modeling/megatron_glue_model.py [hunk 9]
```diff
@@ -0,0 +1,284 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_glue_model.py:1]: 2022

File: nemo/collections/nlp/models/language_modeling/megatron_glue_model.py [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        if not HAVE_APEX:
+            raise ImportError(
+                "Apex was not found. Please see the NeMo README for installation instructions: https://github.com/NVIDIA/NeMo#megatron-gpt."
+            )
+        super().__init__(cfg, trainer)
+        # TODO: Fix this once apex patches FusedScaledMaskedSoftmax.
+        # This is a workaround for the fact that `masked_softmax_fusion` has issues with certain input sizes that may be present while finetuning.
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_glue_model.py:54]: Is this comment still relevant? What needs to be fixed?

File: nemo/collections/nlp/models/language_modeling/megatron_glue_model.py [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+    HAVE_APEX = False
+
+
+__all__ = ['MegatronT5FineTuneModel']
+
+
+class MegatronT5FineTuneModel(NLPModel):
+    """
+    Megatron T5 finetuning
+    """
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_glue_model.py:45]: Please add some more information here about what this class is.

**Review by ericharper — APPROVED** [review]
LGTM. Just a few minor comments. Thanks!

File: nemo/collections/nlp/models/language_modeling/megatron_glue_model.py (hunk 10 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_glue_model.py:54]: Yeah it is still relevant. I need to disable masked_softmax_fusion while fine-tuning because it breaks on certain tensor shapes.
